### PR TITLE
Add a compiled inner func to sync amac

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
See this comment: https://github.com/pytorch-labs/float8_experimental/pull/220#issuecomment-1948923588

I still need to verify that this works, I also want to add a test for when we compile the outer func.. not sure what it happends
cc @awgu, @bdhirsh

### CODE REVIEW

**Comment by drisspg** [comment]
When I log graph breaks I get: 

```
[DEBUG]   File "/home/drisspg/miniconda3/envs/nightly/lib/python3.10/site-packages/torch/_dynamo/exc.py", line 190, in unimplemented
[rank7]:[2024-02-16 10:08:07,178] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG]     raise Unsupported(msg)
[rank7]:[2024-02-16 10:08:07,178] [0/0] torch._dynamo.symbolic_convert.__graph_breaks: [DEBUG] torch._dynamo.exc.Unsupported: call_function args: NestedUserFunctionVariable() ```

**Comment by drisspg** [comment]
Actually perf looks fine

File: test/test_compile.py [hunk 1]
```diff
@@ -201,5 +207,17 @@ def test_float8_graph_output(self):
...
+@unittest.skipIf(not torch.cuda.is_available() or not is_H100, "CUDA not available")
+def test_sync_amax_func():
+    cnts = CompileCounterWithBackend("inductor")
+    module = torch.nn.Sequential(
+        nn.Linear(16, 32, bias=True), nn.ReLU(), nn.Linear(32, 16, bias=True)
+    )
+    float8_mod = swap_linear_with_float8_linear(module, Float8DynamicLinear)
+    compiled_swap_func = torch.compile(sync_float8_amax_and_scale_history, backend=cnts)
+    compiled_swap_func(float8_mod)
+    assert cnts.frame_count == 0, "Compiled graph should have 1 frame!"
```
Comment (awgu) [inline test/test_compile.py:219]: This is not intuitive to me. I was curious why the frame count is not 1?

Comment (drisspg) [reply test/test_compile.py:219]: This is not intuitive to me either and actually this test is not really doing what I thought it was doing, I talked to Brian yesterday. What I think I think is happening is that the outer compile which I am checking sees the inner compiled region and just backtracks out not compiling anything. So this test doesn't really do anything..

File: float8_experimental/float8_linear_utils.py [hunk 2]
```diff
@@ -185,103 +185,110 @@ def sync_float8_amax_and_scale_history(model: torch.nn.Module, fp8_layers=None)
...
+            child.fp8_scale_dL_dY.copy_(new_dL_dY_scales[idx])
 
-    # Iterate through the layers and update the scales, and set the flag to signal that the amaxes/scales are ready
-    for idx, child in enumerate(fp8_layers):
-        child.fp8_scale_x.copy_(new_x_scales[idx])
-        child.fp8_scale_w.copy_(new_w_scales[idx])
-        child.fp8_scale_dL_dY.copy_(new_dL_dY_scales[idx])
+    # This allows for the compile to succede on the inner func and fail on the graph breaks
+    # at the beginning and and of syncing
+    inner_func()
```
Comment (bdhirsh) [inline float8_experimental/float8_linear_utils.py:289]: a little hard to tell but I think this is all code motion? (moving it inside of inner_func())

File: float8_experimental/float8_linear_utils.py [hunk 3]
```diff
@@ -185,103 +185,110 @@ def sync_float8_amax_and_scale_history(model: torch.nn.Module, fp8_layers=None)
...
-        x_dtypes.add(child.last_seen_input_dtype)
-        scale_fn_recipes.add(child.recipe.scale_fn_name)
-
-    # TODO This way to get the activation dtype is not ideal
-    if len(x_dtypes) != 1:
-        raise ValueError(
-            f"All layers must have the same last seen input_dtype, got {x_dtypes}"
-        )
-    x_dtype = next(iter(x_dtypes))
+    def inner_func():
```
Comment (bdhirsh) [inline float8_experimental/float8_linear_utils.py:188]: maybe add a comment on why we want the inner fn? (we have graph breaks that we want to fix eventually, but in the meantime we want to ensure that everything in inner_func is compiled)

**Review by bdhirsh — APPROVED** [review]
stamp

File: float8_experimental/float8_linear_utils.py (hunk 2 above)
Comment (drisspg) [reply float8_experimental/float8_linear_utils.py:289]: yeah tabs strike again here, but just indented one layer

File: float8_experimental/float8_linear_utils.py [hunk 4]
```diff
@@ -185,103 +185,110 @@ def sync_float8_amax_and_scale_history(model: torch.nn.Module, fp8_layers=None)
...
-        child.fp8_scale_w.copy_(new_w_scales[idx])
-        child.fp8_scale_dL_dY.copy_(new_dL_dY_scales[idx])
+    # This allows for the compile to succede on the inner func and fail on the graph breaks
+    # at the beginning and and of syncing
+    inner_func()
 
+    for child in fp8_layers:
         # 4. set a flag to signal amaxes/scales are ready
         # We only update the flag if we know it will be checked by the modules
-        if fp8_config.enable_amax_init:
```
[the remark this answers is not in the document]
Comment (drisspg) [reply float8_experimental/float8_linear_utils.py:286]: This was andrews original comment here: https://github.com/pytorch-labs/float8_experimental/pull/220#issuecomment-1948923588

I used the wrong config setting

**Pull request closed without merging**